### ISSUE TITLE
Package info page not working with UpstreamError.

### ISSUE THREAD

**Issue by turokm** [devpi/devpi#759]
I'm having a bit of an issue with devpi's web interface.

I'm running these versions:
```
$ pip list | grep devpi
devpi-client        5.1.0              
devpi-common        3.4.0              
devpi-server        5.2.0              
devpi-web           4.0.0              
```

I started a devpi server, and used the devpi client to upload several *.tgz files.

They were registered and uploaded:

```
$ devpi upload *
register foo-0.5.0.4 to http://localhost:8080/testuser/dev/
file_upload of foo-0.5.0.4.tar.gz to http://localhost:8080/testuser/dev/
register foo-0.5 to http://localhost:8080/testuser/dev/
file_upload of foo-0.5.tar.gz to http://localhost:8080/testuser/dev/
```

When I navigate to the web page for http://localhost:8080/testuser/dev, I only see version 0.5.0.4.

I can download the .tgz from the link under "Releases", but the "Info page" link shows:

```
Internal Server Error

The server encountered an unexpected internal server error

(generated by waitress)
```

And the server log says:
```
2019-12-18 16:09:24,303 ERROR Exception while serving /testuser/dev/foo/0.5.0.4
Traceback (most recent call last):
  File "/home/qestrat1/system/conda/envs/pypiserver-local/lib/python3.8/site-packages/pyramid/tweens.py", line 13, in _error_handler
    response = request.invoke_exception_view(exc_info)
  File "/home/qestrat1/system/conda/envs/pypiserver-local/lib/python3.8/site-packages/pyramid/view.py", line 779, in invoke_exception_view
    raise HTTPNotFound
pyramid.httpexceptions.HTTPNotFound: The resource could not be found.
```

Not sure if I'm missing something simple and obvious.

**Comment by fschulze** [comment]
Try http://localhost:8080/testuser/dev/foo to see the pkg overview page, it should list the other version(s).

Is the above exception the only exception? Could you paste some more lines before and after the exception to get more context? Also try running devpi-server with ``--debug`` to get more log output.

**Comment by turokm** [comment]
Thank you.   I realize I should mention that I'm behind a firewall without having set my proxy.

I restarted the server with --debug, and it seems like there is an exception being thrown when it also tries to go to pypi.org for this package (which only should exist locally).

You can see the attempt to access the tar.gz, and that was successful.

This first set is trying to retrieve the info page, and below, getting the url at:  http://localhost:8080/testuser/dev/foo

```
2019-12-19 14:06:30,721 INFO  [req21] GET /
2019-12-19 14:06:30,732 DEBUG [req21] [Rtx143] in-transaction 143
2019-12-19 14:06:30,761 DEBUG [req21] [Rtx143] closing transaction at 143
2019-12-19 14:06:30,762 DEBUG [req21] 200 OK 0.041s serial=143 length=6278 type=text/html; charset=UTF-8
2019-12-19 14:06:36,902 INFO  [req22] GET /testuser/dev
2019-12-19 14:06:36,916 DEBUG [req22] [Rtx143] in-transaction 143
2019-12-19 14:06:36,977 DEBUG [req22] [Rtx143] closing transaction at 143
2019-12-19 14:06:36,977 DEBUG [req22] 200 OK 0.075s serial=143 length=7907 type=text/html; charset=UTF-8
2019-12-19 14:06:40,721 INFO  [req23] GET /
2019-12-19 14:06:40,734 DEBUG [req23] [Rtx143] in-transaction 143
2019-12-19 14:06:40,768 DEBUG [req23] [Rtx143] closing transaction at 143
2019-12-19 14:06:40,768 DEBUG [req23] 200 OK 0.047s serial=143 length=6278 type=text/html; charset=UTF-8
2019-12-19 14:06:49,671 INFO  [req24] GET /testuser/dev/+f/1ca/39ea6c9e54c4d/foo-0.5.0.4.tar.gz
2019-12-19 14:06:49,696 DEBUG [req24] [Rtx143] in-transaction 143
2019-12-19 14:06:49,723 DEBUG [req24] [Rtx143] closing transaction at 143
2019-12-19 14:06:49,723 DEBUG [req24] 200 OK 0.052s serial=143 length=42167 type=application/x-tar
2019-12-19 14:06:50,721 INFO  [req25] GET /
2019-12-19 14:06:50,735 DEBUG [req25] [Rtx143] in-transaction 143
2019-12-19 14:06:50,768 DEBUG [req25] [Rtx143] closing transaction at 143
2019-12-19 14:06:50,768 DEBUG [req25] 200 OK 0.047s serial=143 length=6278 type=text/html; charset=UTF-8
2019-12-19 14:06:54,734 INFO  [req26] GET /testuser/dev/foo/0.5.0.4
2019-12-19 14:06:54,747 DEBUG [req26] [Rtx143] in-transaction 143
2019-12-19 14:06:54,788 DEBUG [req26] [Rtx143] reading index https://pypi.org/simple/foo/
2019-12-19 14:06:54,790 DEBUG Starting new HTTPS connection (4): pypi.org:443
2019-12-19 14:06:55,048 WARNI [req26] [Rtx143] OS error during httpget of https://pypi.org/simple/foo/ at /home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/extpypi.py:392::get_simplelinks_perstage
2019-12-19 14:06:55,049 DEBUG [req26] [Rtx143] transaction rollback at 143
2019-12-19 14:06:55,049 DEBUG [req26] [Rtx143] closing transaction at 143
2019-12-19 14:06:55,049 ERROR Exception while serving /testuser/dev/foo/0.5.0.4
Traceback (most recent call last):
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/tweens.py", line 13, in _error_handler
    response = request.invoke_exception_view(exc_info)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/view.py", line 779, in invoke_exception_view
    raise HTTPNotFound
pyramid.httpexceptions.HTTPNotFound: The resource could not be found.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/waitress/channel.py", line 356, in service
    task.service()
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/waitress/task.py", line 172, in service
    self.execute()
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/waitress/task.py", line 440, in execute
    app_iter = self.channel.server.application(environ, start_response)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/views.py", line 153, in __call__
    return self.app(environ, start_response)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 270, in __call__
    response = self.execution_policy(environ, self)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 279, in default_execution_policy
    return request.invoke_exception_view(reraise=True)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/view.py", line 778, in invoke_exception_view
    reraise_(*exc_info)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/compat.py", line 179, in reraise
    raise value
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 277, in default_execution_policy
    return router.invoke_request(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 249, in invoke_request
    response = handle_request(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/views.py", line 177, in request_log_handler
    response = handler(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/views.py", line 213, in request_tx_handler
    response = handler(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/tweens.py", line 43, in excview_tween
    response = _error_handler(request, exc)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/tweens.py", line 17, in _error_handler
    reraise(*exc_info)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/compat.py", line 179, in reraise
    raise value
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/tweens.py", line 41, in excview_tween
    response = handler(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 147, in handle_request
    response = _call_view(
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/view.py", line 667, in _call_view
    response = view_callable(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/config/views.py", line 169, in __call__
    return view(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/config/views.py", line 188, in attr_view
    return view(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/config/views.py", line 214, in predicate_wrapper
    return view(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/viewderivers.py", line 436, in rendered_view
    result = view(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_web/views.py", line 629, in version_get
    whitelist_info = stage.get_mirror_whitelist_info(name)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/model.py", line 772, in get_mirror_whitelist_info
    in_index = stage.has_project_perstage(project)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/extpypi.py", line 501, in has_project_perstage
    self.get_simplelinks_perstage(project)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/extpypi.py", line 408, in get_simplelinks_perstage
    raise self.UpstreamError("%s status on GET %s" % (
devpi_server.model.UpstreamError: -1 status on GET https://pypi.org/simple/foo/
2019-12-19 14:07:00,721 INFO  [req27] GET /
2019-12-19 14:07:00,735 DEBUG [req27] [Rtx143] in-transaction 143
2019-12-19 14:07:00,767 DEBUG [req27] [Rtx143] closing transaction at 143
2019-12-19 14:07:00,767 DEBUG [req27] 200 OK 0.046s serial=143 length=6278 type=text/html; charset=UTF-8
```

And the access of the specific path:   http://localhost:8080/testuser/dev/foo

```
2019-12-19 14:10:31,387 INFO  [req50] GET /testuser/dev/foo/
2019-12-19 14:10:31,402 DEBUG [req50] [Rtx143] in-transaction 143
2019-12-19 14:10:31,430 DEBUG [req50] [Rtx143] get_simplelinks_perstage: private package 'foo' not whitelisted, ignoring root/pypi
2019-12-19 14:10:31,661 DEBUG [req50] [Rtx143] reading index https://pypi.org/simple/foo/
2019-12-19 14:10:31,663 DEBUG Starting new HTTPS connection (6): pypi.org:443
2019-12-19 14:10:31,674 WARNI [req50] [Rtx143] OS error during httpget of https://pypi.org/simple/foo/ at /home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/extpypi.py:392::get_simplelinks_perstage
2019-12-19 14:10:31,674 DEBUG [req50] [Rtx143] transaction rollback at 143
2019-12-19 14:10:31,674 DEBUG [req50] [Rtx143] closing transaction at 143
2019-12-19 14:10:31,674 ERROR Exception while serving /testuser/dev/foo/
Traceback (most recent call last):
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/tweens.py", line 13, in _error_handler
    response = request.invoke_exception_view(exc_info)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/view.py", line 779, in invoke_exception_view
    raise HTTPNotFound
pyramid.httpexceptions.HTTPNotFound: The resource could not be found.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/waitress/channel.py", line 356, in service
    task.service()
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/waitress/task.py", line 172, in service
    self.execute()
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/waitress/task.py", line 440, in execute
    app_iter = self.channel.server.application(environ, start_response)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/views.py", line 153, in __call__
    return self.app(environ, start_response)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 270, in __call__
    response = self.execution_policy(environ, self)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 279, in default_execution_policy
    return request.invoke_exception_view(reraise=True)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/view.py", line 778, in invoke_exception_view
    reraise_(*exc_info)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/compat.py", line 179, in reraise
    raise value
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 277, in default_execution_policy
    return router.invoke_request(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 249, in invoke_request
    response = handle_request(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/views.py", line 177, in request_log_handler
    response = handler(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/views.py", line 213, in request_tx_handler
    response = handler(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/tweens.py", line 43, in excview_tween
    response = _error_handler(request, exc)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/tweens.py", line 17, in _error_handler
    reraise(*exc_info)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/compat.py", line 179, in reraise
    raise value
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/tweens.py", line 41, in excview_tween
    response = handler(request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/router.py", line 147, in handle_request
    response = _call_view(
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/view.py", line 667, in _call_view
    response = view_callable(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/config/views.py", line 169, in __call__
    return view(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/config/views.py", line 188, in attr_view
    return view(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/config/views.py", line 214, in predicate_wrapper
    return view(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/pyramid/viewderivers.py", line 436, in rendered_view
    result = view(context, request)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_web/views.py", line 556, in project_get
    whitelist_info = context.stage.get_mirror_whitelist_info(context.project)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/model.py", line 772, in get_mirror_whitelist_info
    in_index = stage.has_project_perstage(project)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/extpypi.py", line 501, in has_project_perstage
    self.get_simplelinks_perstage(project)
  File "/home/qestrat1/system/conda/envs/devpi/lib/python3.8/site-packages/devpi_server/extpypi.py", line 408, in get_simplelinks_perstage
    raise self.UpstreamError("%s status on GET %s" % (
devpi_server.model.UpstreamError: -1 status on GET https://pypi.org/simple/foo/
2019-12-19 14:10:40,721 INFO  [req51] GET /
2019-12-19 14:10:40,739 DEBUG [req51] [Rtx143] in-transaction 143
2019-12-19 14:10:40,769 DEBUG [req51] [Rtx143] closing transaction at 143
2019-12-19 14:10:40,770 DEBUG [req51] 200 OK 0.049s serial=143 length=6278 type=text/html; charset=UTF-8
```

**Comment by turokm** [comment]
So I set my proxy in the process where I was running the devpi server, and now the pages at http://localhost:8080/testuser/dev/foo works, as does the specific version info page at http://localhost:8080/testuser/dev/foo/0.5.0.4.

Is there some setting to configure this behavior?

**Comment by fschulze** [comment]
This should be fixed in the upcoming devpi-server 6.0.0 by #794, also see #451
